assert: adjust loose assertions

BridgeAR opened the pull request:
This changes the loose deep equal comparison by using the same logic
as done in the strict deep equal comparison besides comparing
primitives loosely, not comparing symbol properties and not comparing
the prototype.

`assert.deepEqual` is still commenly used and this is likely the
biggest pitfall.

Most changes are only minor and won't have a big impact besides
likely fixing user expectations.

I guess this might still be a bit bold but I believe it's the right
thing to do. We just have to check how much impact this has in
userland.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/19474/
CITGM https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1683/

BridgeAR commented:
Rebased due to conflicts.

@nodejs/tsc PTAL

CI https://ci.nodejs.org/job/node-test-pull-request/19615/

Trott commented:
Failure on ARM appears Jenkins/Java related. Resume Build CI: https://ci.nodejs.org/job/node-test-pull-request/19633/ (scheduled)

BridgeAR commented:
@nodejs/tsc PTAL. This needs another review due to being semver-major.

BridgeAR commented:
@nodejs/tsc PTAL

BridgeAR commented:
@nodejs/tsc PTAL. This needs some reviews.

BridgeAR commented:
Rebased due to conflicts.

CI https://ci.nodejs.org/job/node-test-pull-request/19858/

@nodejs/tsc please have a look. The PR is semver-major and is open for quite some while without getting a second review.

BridgeAR commented:
@nodejs/tsc PTAL.

This PR is open for quite some while and needs some reviews due to being semver-major.

danbev commented on test/parallel/test-assert-deep.js:
```diff
@@ -157,28 +157,26 @@ assert.throws(
     new Int8Array([1]), // Int8Array
-    new Uint8Array([1]), // Uint8Array
+    // TODO: Fix this check...
+    // new Uint8Array([1]), // Uint8Array
```
It looks like this check is commented out on purpose, perhaps for a later PR, but I just wanted to double check if that is the case?

BridgeAR replied:
`Buffer` is an `Uint8Array` and without checking the prototype, this is identical. I guess just removing the entry would be best.

mhdawson commented:
Do you have any data for:
>I guess this might still be a bit bold but I believe it's the right
thing to do. We just have to check how much impact this has in
userland.

gabrielschulhof commented on doc/api/assert.md:
```diff
@@ -163,6 +163,10 @@ An alias of [`assert.ok()`][].
 changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/REPLACEME
+    description: The type tags are now properly compared and a couple minor
```
Nit: Perhaps change the wording to "... and *there are* a couple *of* minor ..."

gabrielschulhof commented on doc/api/assert.md:
```diff
@@ -191,26 +195,39 @@ An alias of [`assert.deepStrictEqual()`][].
+potentially surprising results.
+
+"Deep" equality means that the enumerable "own" properties of child objects
+are recursively evaluated also by the following rules.
```
Nit: "... are  *also* recursively evaluated ~~also~~ by the ... "

BridgeAR commented:
CITGM https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1715/
CI https://ci.nodejs.org/job/node-test-pull-request/20127/

@mhdawson I just started CITGM. That is likely the best way to check this.

BridgeAR commented:
The CITGM run looks good to me. I could not find a single failure related to this change.

ChALkeR commented on doc/api/assert.md:
```diff
@@ -1249,10 +1266,8 @@ second argument. This might lead to difficult-to-spot errors.
 [`ERR_INVALID_RETURN_VALUE`]: errors.html#errors_err_invalid_return_value
 [`Error.captureStackTrace`]: errors.html#errors_error_capturestacktrace_targetobject_constructoropt
 [`Error`]: errors.html#errors_class_error
-[`Map`]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map
```
Why is this being removed? `Map` and `Set` are mentioned above.

BridgeAR replied:
I removed the only two entries which were linked. Shall I change the existing entries to add the link?

BridgeAR replied:
I referenced all entries now.

ChALkeR commented on doc/api/assert.md:
```diff
@@ -1249,10 +1266,8 @@ second argument. This might lead to difficult-to-spot errors.
-[`Map`]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map
 [`Object.is()`]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is
 [`RegExp`]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions
-[`Set`]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set
```
Ditto.

BridgeAR replied:
See above.

BridgeAR commented:
CI Lite https://ci.nodejs.org/job/node-test-pull-request-lite-pipeline/2266/ (I just used the lite CI since the full CI was full and there's just a minor documentation change since the last full CI).

ChALkeR commented:
Are any of the CITGM failures related to this change?

BridgeAR commented:
@ChALkeR multiple failures are issues to run the tests on Windows. More failures are related to C++ API removals from V8 and we already track most failures in #25060 and a couple are related to a version. There are a couple further issues besides these that are also unrelated. I could not find any related issue.

BridgeAR commented:
I'll go ahead and land this if there are no concerns up until the 22.01.

BridgeAR commented:
Landed in 7493db21b667ed746d39c9b54357eac4287232e3 :tada:

Thanks everyone for the reviews.
